src/org/zaproxy/zap/extension/cspscanner/ContentSecurityPolicyScanner.java

```diff
@@ -0,0 +1,310 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ * 
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ * 
+ * Copyright 2017 The ZAP Development Team
+ *  
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License. 
+ * You may obtain a copy of the License at 
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0 
+ *   
+ * Unless required by applicable law or agreed to in writing, software 
+ * distributed under the License is distributed on an "AS IS" BASIS, 
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
+ * See the License for the specific language governing permissions and 
+ * limitations under the License. 
+ */
+
+package org.zaproxy.zap.extension.cspscanner;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Vector;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import net.htmlparser.jericho.Source;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Plugin.AlertThreshold;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PassiveScanThread;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+import com.shapesecurity.salvation.ParserWithLocation;
+import com.shapesecurity.salvation.data.Notice;
+import com.shapesecurity.salvation.data.Origin;
+import com.shapesecurity.salvation.data.Policy;
+import com.shapesecurity.salvation.data.URI;
+
+/**
+ * Content Security Policy Header passive scan rule
+ * https://github.com/zaproxy/zaproxy/issues/527
+ * Meant to compliment the CSP Header Missing passive scan rule
+ * 
+ * @author kingthorin+owaspzap@gmail.com
+ * TODO: Add handling for multiple CSP headers 
+ * TODO: Add handling for CSP via META tag 
+ * See https://github.com/shapesecurity/salvation/issues/149
+ * for info on combining CSP policies
+ */
+public class ContentSecurityPolicyScanner extends PluginPassiveScanner {
+
+	private static final String MESSAGE_PREFIX = "cspscanner.";
+	private static final int PLUGIN_ID = 10055;
+	private static final Logger LOGGER = Logger.getLogger(ContentSecurityPolicyScanner.class);
+
+	private static final String HTTP_HEADER_CSP = "Content-Security-Policy";
+	private static final String HTTP_HEADER_XCSP = "X-Content-Security-Policy";
+	private static final String HTTP_HEADER_WEBKIT_CSP = "X-WebKit-CSP";
+
+	private static final String WILDCARD_URI="http://*";
+	
+	private boolean cspHeaderFound = false;

```

Comment:
This could be moved to the method `scanHttpResponseReceive`, just used there, it also prevents a `NullPointerException` when scanning consecutive messages, the first with CSP header and the following without.